packages/autorun-luajit/src/types/common.rs

```diff
@@ -0,0 +1,427 @@
+pub const LJ_TNUMX: u32 = !13u32;
+
+pub trait IntoLJType {
+	const LJ_TYPE: u32;
+}
+
+pub type MSize = u64;
+pub type GCSize = u64;
+
+#[repr(C)]
+pub struct lua_Debug {
+	pub event: i32,
+	pub name: *const u8,
+	pub namewhat: *const u8,
+	pub what: *const u8,
+	pub source: *const u8,
+	pub currentline: i32,
+	pub nups: i32,
+	pub linedefined: i32,
+	pub lastlinedefined: i32,
+	pub short_src: [u8; LUA_IDSIZE as usize],
+	pub i_ci: i32,
+}
+
+pub type lua_Hook = extern "C-unwind" fn(L: *mut lua_State, ar: *mut lua_Debug);
+pub type lua_CFunction = extern "C-unwind" fn(L: *mut lua_State) -> i32;
+pub type lua_Alloc = extern "C-unwind" fn(
+	ud: *mut core::ffi::c_void,
+	ptr: *mut core::ffi::c_void,
+	osize: usize,
+	nsize: usize,
+) -> *mut core::ffi::c_void;
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct MRef {
+	pub ptr64: u64,
+}
+
+impl MRef {
+	// equivalent to the mref macro in LuaJIT
+	pub fn as_ptr<T>(&self) -> *mut T {
+		self.ptr64 as *mut T
+	}
+
+	pub fn tvref(&self) -> *mut TValue {
+		self.as_ptr::<TValue>()
+	}
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct GCRef {
+	pub gcptr64: u64,
+}
+
+impl GCRef {
+	// equivalent to the gcref macro in LuaJIT
+	pub fn as_ptr<T>(&self) -> *mut T {
+		self.gcptr64 as *mut T
+	}
+}
+
+#[repr(C, packed)]
+#[derive(Clone, Copy, Debug)]
+pub struct GCHeader {
+	pub nextgc: GCRef,
+	pub marked: u8,
+	pub gct: u8,
+}
+
+// #define LJ_GCVMASK		(((uint64_t)1 << 47) - 1)
+pub const LJ_GCVMASK: u64 = ((1u64 << 47) - 1);
+
+#[repr(C, align(8))]
+#[derive(Clone, Copy)]
+pub union TValue {
+	pub u64: u64,
+	pub f64: f64,
+	pub gcr: GCRef,
+	pub it64: i64,
+}
+
+impl TValue {
+	pub fn as_ptr<T: IntoLJType>(&self) -> anyhow::Result<*mut T> {
+		if self.itype() != T::LJ_TYPE {
+			anyhow::bail!("TValue type mismatch: expected {}, got {}", T::LJ_TYPE, self.itype());
+		}
+
+		Ok(unsafe { (self.gcr.gcptr64 & LJ_GCVMASK) as *mut T })
+	}
+
+	pub fn as_ref<T: IntoLJType>(&self) -> anyhow::Result<&T> {
+		Ok(unsafe { &*self.as_ptr::<T>()? })
+	}
+
+	pub fn as_mut<T: IntoLJType>(&mut self) -> anyhow::Result<&mut T> {
+		Ok(unsafe { &mut *self.as_ptr::<T>()? })
+	}
+
+	pub fn itype(&self) -> u32 {
+		unsafe { ((self.it64 >> 47) & 0xFFFFFFFF) as u32 }
+	}
+}
+
+#[repr(C, packed)]
+#[derive(Clone, Copy, Debug)]
+pub struct GCFuncHeader {
+	pub header: GCHeader,
+	pub ffid: u8,
+	pub nupvalues: u8,
+	pub env: GCRef,
+	pub gclist: GCRef,
+	pub pc: MRef,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct GCfuncC {
+	pub header: GCFuncHeader,
+	pub c: lua_CFunction,
+	pub upvalue: [TValue; 1],
+}
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct GCfuncL {
+	pub header: GCFuncHeader,
+	pub uvptr: [GCRef; 1],
+}
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub union GCfunc {
+	pub c: GCfuncC,
+	pub l: GCfuncL,
+}
+
+impl IntoLJType for GCfunc {
+	const LJ_TYPE: u32 = LJ_TFUNC;
+}
+
+pub const FF_LUA: u8 = 0;
+pub const FF_C: u8 = 1;
+
+impl GCfunc {
+	pub fn as_l(&self) -> &GCfuncL {
```

Comment:
Shouldn't just expose actually unsafe behavior in a seemingly safe api.

Safer way to do this seems to be to add pub fn header(&self) -> &GCFuncHeader {}, use that in all the methods, and make as_lua and as_c return Option\<&GCFuncL\> / Option\<&GCFuncC\>